frontend/src/scenes/surveys/surveyActivityDescriber.tsx

```diff
@@ -410,3 +440,177 @@ function getPreposition(field: string): string {
             return 'of'
     }
 }
+
+type SurveyQuestion = BasicSurveyQuestion | LinkSurveyQuestion | RatingSurveyQuestion | MultipleSurveyQuestion
+
+export function describeQuestionChanges(before: SurveyQuestion, after: SurveyQuestion): JSX.Element[] {
+    const commonChanges = describeCommonChanges(before, after)
+    const typeChangeDescription =
+        before.type !== after.type
+            ? [
+                  <>
+                      changed question type from <strong>{before.type}</strong> to <strong>{after.type}</strong>
+                  </>,
+              ]
+            : []
+
+    const specificChanges = match([before, after] as const)
+        .with([{ type: SurveyQuestionType.Link }, { type: SurveyQuestionType.Link }], describeLinkChanges)
+        .with([{ type: SurveyQuestionType.Rating }, { type: SurveyQuestionType.Rating }], describeRatingChanges)
+        .with(
+            [
+                { type: P.union(SurveyQuestionType.SingleChoice, SurveyQuestionType.MultipleChoice) },
+                { type: P.union(SurveyQuestionType.SingleChoice, SurveyQuestionType.MultipleChoice) },
+            ],
+            describeMultipleChoiceChanges
+        )
+        .otherwise(() => [])
+
+    return [...commonChanges, ...typeChangeDescription, ...specificChanges, ...describeBranchingChanges(before, after)]
+}
+
+export function describeCommonChanges(before: SurveyQuestion, after: SurveyQuestion): JSX.Element[] {
```

Comment:
I still have functions that are comparing outcomes conditionally, but now that they can be flattened by the `.with()` combinator, this whole chunk of code gets easier to read.